### PR TITLE
lightning.data: Remove torch distributed for the Dataset Optimizer

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR replaces the current mechanism for syncing some objects across machines with torch.distributed by a server based version. This avoids locking the machine on start until they are all available.

Fixes #\<issue_number>

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--19182.org.readthedocs.build/en/19182/

cc @borda

### CODE REVIEW

**Review by rusenask — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

File: src/lightning/data/utilities/broadcast.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+        token = None
+        if os.getenv("LIGHTNING_CLOUD_URL") is not None:
+            payload = {"apiKey": os.getenv("LIGHTNING_API_KEY"), "username": os.getenv("LIGHTNING_USERNAME")}
+            url_login = os.getenv("LIGHTNING_CLOUD_URL", "") + "/v1/auth/login"
+            res = requests.post(url_login, data=json.dumps(payload))
+            if "token" not in res.json():
+                raise RuntimeError(
+                    f"You haven't properly setup your environment variables with {url_login} and data: \n{payload}"
+                )
+            token = res.json()["token"]
```
Comment (awaelchli) [inline src/lightning/data/utilities/broadcast.py:154]: This should probably be a function/method? Then you don't need a comment "get the token"

File: src/lightning/data/utilities/broadcast.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    ) -> Any:
+        url = urljoin(self.base_url, path)
+        return self.session.post(url, data=data, params=query_params, json=json)
+
+    @_http_method_logger_wrapper
+    def delete(self, path: str) -> Any:
+        url = urljoin(self.base_url, path)
+        return self.session.delete(url)
+
+    def log_function(self, message: str, *args: Any, **kwargs: Any) -> None:
```
Comment (awaelchli) [inline src/lightning/data/utilities/broadcast.py:126]: What is the example for it though? Are we going to use it ourselves?

**Review by awaelchli — APPROVED** [review]

File: src/lightning/data/utilities/broadcast.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
+# Copyright The Lightning AI team.
```
[the remark this answers is not in the document]
Comment (tchaton) [reply src/lightning/data/utilities/broadcast.py:1]: sure.

File: src/lightning/data/utilities/broadcast.py (hunk 2 above)
Comment (tchaton) [reply src/lightning/data/utilities/broadcast.py]: This isn't needed. I removed it.

File: src/lightning/data/utilities/broadcast.py (hunk 1 above)
Comment (tchaton) [reply src/lightning/data/utilities/broadcast.py]: sure.

**Pull request merged**